### PR TITLE
[FEATURE] Introduce Contents directive

### PR DESCRIPTION
**Pull request by linawolf**

### CODE REVIEW

File: lib/Directives/Contents.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace Doctrine\RST\Directives;
+
+use Doctrine\RST\Nodes\Node;
+use Doctrine\RST\Parser;
+
+final class Contents extends Directive
+{
+    public function __construct()
+    {
+    }
```
Comment (greg0ire) [inline lib/Directives/Contents.php]: :thinking: what is the point of that empty constructor?

File: lib/NodeFactory/NodeFactory.php [hunk 2]
```diff
@@ -45,6 +46,9 @@ public function createDocumentNode(Environment $environment): DocumentNode;
      */
     public function createTocNode(Environment $environment, array $files, array $options): TocNode;
 
+    /** @param string[] $options */
+    public function createContentsNode(Environment $environment, DocumentNode $documentNode, array $options): ContentsNode;
```
Comment (greg0ire) [inline lib/NodeFactory/NodeFactory.php:50]: This is technically a BC-break. A solution would be to document this with `@method` on the interface, then switch from `@method` to a real method signature on 0.6.x, or to target 0.6.x directly. I think the latter solution would work be best, it's more straightforward and we don't need that level of user-friendliness for that library.

File: lib/Nodes/ContentsNode.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\RST\Nodes;
+
+use Doctrine\RST\Environment;
+
+class ContentsNode extends Node
```
Comment (greg0ire) [inline lib/Nodes/ContentsNode.php:9]: Should this class become `final`?

File: lib/Directives/Contents.php (hunk 1 above)
Comment (linawolf) [reply lib/Directives/Contents.php]: Leftover. Removed

File: lib/Nodes/ContentsNode.php (hunk 3 above)
Comment (linawolf) [reply lib/Nodes/ContentsNode.php:9]: The other nodes are not final so I figured they are not final by design?

File: lib/NodeFactory/NodeFactory.php (hunk 2 above)
Comment (linawolf) [reply lib/NodeFactory/NodeFactory.php:50]: To be honest I don't get this whole construct with the Interface? This makes adding any renderer a breaking change? but I don't mind to target 0.6.x

Comment (greg0ire) [reply lib/NodeFactory/NodeFactory.php:50]: Yeah there's only one implementation, and having that big of an interface makes it near impossible to implement… the reasoning behind it is here: https://github.com/doctrine/rst-parser/pull/13#discussion_r221410489

@jwage , I see you have a use case for this, do you actually have a custom implementation of this? And if you do use wrapping, do you have a proxy method for each and every method in this interface?

File: lib/Nodes/ContentsNode.php (hunk 3 above)
Comment (greg0ire) [reply lib/Nodes/ContentsNode.php:9]: Let's suppose so.

**Comment by linawolf** [comment]
@greg0ire So I changed the base branch to 0.6.x, how shall we proceed? can this be reviewed and then merged? Or is there like a certain date where development is started for version 0.6?

**Comment by greg0ire** [comment]
There is no time constraint, the branch exists, so it's open for contributions :slightly_smiling_face:

File: lib/Utility/TitleLinkUtility.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    /** @param string[]|string $title */
+    private function generateAnchorFromTitle($title): string
+    {
+        $slug = is_array($title)
+            ? $title[1]
+            : $title;
+
+        return Environment::slugify($slug);
+    }
+}
```
Comment (greg0ire) [inline lib/Utility/TitleLinkUtility.php:105]: Have you considered covering the lines codecov points at with tests?

Comment (linawolf) [reply lib/Utility/TitleLinkUtility.php:105]: I didn't introduce the code, just moved it, but sure, Y not?

Comment (greg0ire) [reply lib/Utility/TitleLinkUtility.php:105]: Oh sorry, didn't notice that. If you can that's great, but it's not a requirement.

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @linawolf !

**Pull request merged**